File: VITS/src/main/java/com/chatwaifu/vits/utils/FileUtils.java
/*
 * Copyright (c) 2013, David Brodsky. All rights reserved.
 *
 *	This program is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *	
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *	
 *	You should have received a copy of the GNU General Public License
 *	along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.chatwaifu.vits.utils;

import android.content.res.AssetManager;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;


public class FileUtils {

    static final String TAG = "FileUtils";
//
//    public static void copy(File src, File dst) throws IOException {
//        InputStream in = new FileInputStream(src);
//        OutputStream out = new FileOutputStream(dst);
//
//        // Transfer bytes from in to out
//        byte[] buf = new byte[1024];
//        int len;
//        while ((len = in.read(buf)) > 0) {
//            out.write(buf, 0, len);
//        }
//        in.close();
//        out.close();
//    }
//
//    /**
//     * Delete a directory and all its contents
//     */
//    public static void deleteDirectory(File fileOrDirectory) {
//        File[] children = fileOrDirectory.listFiles();
//        if (fileOrDirectory.isDirectory() && children != null) {
//            for (File child : children) {
//                deleteDirectory(child);
//            }
//        }
//        fileOrDirectory.delete();
//    }
//
//    /**
//     * 递归拷贝Asset目录中的文件到rootDir中
//     * Recursively copy the files in the Asset directory to rootDir
//     * @param assets
//     * @param path
//     * @param rootDir
//     * @throws IOException
//     */
//    public static void copyAssets(AssetManager assets, String path, String rootDir, byte[] buffer) throws IOException {
//        if (isAssetsDir(assets, path)) {
//            File dir = new File(rootDir + File.separator + path);
//            if (!dir.exists() && !dir.mkdirs()) {
//                throw new IllegalStateException("mkdir failed");
//            }
//            for (String s : assets.list(path)) {
//                copyAssets(assets, path + "/" + s, rootDir, buffer);
//            }
//        } else {
//            InputStream input = assets.open(path);
//            File dest = new File(rootDir, path);
//            copyToFileOrThrow(input, dest, buffer);
//        }
//    }
//
//    public static boolean isAssetsDir(AssetManager assets, String path) {
//        try {
//
//            String[] files = assets.list(path);
//            return files != null && files.length > 0;
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
//        return false;
//    }
//
//    public static int DEFAULT_BUFFER = 64 * 1024;
//    public static void copyToFileOrThrow(InputStream inputStream, File destFile, byte[] buffer)
//            throws IOException {
//        if (destFile.exists()) {
//            return;
//
//        }
//        File file = destFile.getParentFile();
//        if (file != null && !file.exists()) {
//            file.mkdirs();
//        }
//        FileOutputStream out = new FileOutputStream(destFile);
//        try {
//            if(buffer == null || buffer.length == 0) {
//                buffer = new byte[DEFAULT_BUFFER];
//            }
//            // byte[] buffer = new byte[BUFFER];
//            int bytesRead;
//            while ((bytesRead = inputStream.read(buffer)) >= 0) {
//                out.write(buffer, 0, bytesRead);
//            }
//        } finally {
//            out.flush();
//            try {
//                out.getFD().sync();
//            } catch (IOException e) {
//            }
//            out.close();
//        }
//    }
}

File: VITS/src/main/java/com/chatwaifu/vits/utils/file/FileUtils.kt
package com.chatwaifu.vits.utils.file

import android.content.res.AssetManager
import android.util.Log
import com.chatwaifu.vits.data.Bv2Config
import com.google.gson.Gson
import java.io.*


object FileUtils {

    /**
     * 解析并校验 Bert-VITS2 的 config.json。
     *
     * 和老 VITS 的 config 不是一回事：老的靠 `symbols` + `text_cleaners` 驱动我们自己写的
     * 音素化（`ChineseTextUtils` / `JapaneseTextUtils`，已随 ncnn 引擎一起删掉），
     * BV2 的 G2P 在 `text-preprocess` 里做，config 只需要提供两样东西——
     * 采样率（要拿去配 AudioTrack）和 `spk2id`（多人模型的 speaker 表）。
     *
     * 失败原因回给调用方而不是在这里 Toast：模型导入要把它写进 ImportError.InvalidVitsConfig。
     */
    fun parseConfig(path: String): ConfigParseResult {
        return try {
            val configBuffer = File(path).bufferedReader().use { it.readText() }
            val config = Gson().fromJson(configBuffer, Bv2Config::class.java)
                ?: return ConfigParseResult.Failure("config.json 解析为空")
            val data = config.data
                ?: return ConfigParseResult.Failure("config.json 缺少 data 段")
            if (data.sampling_rate == null || data.sampling_rate <= 0) {
                return ConfigParseResult.Failure("config.json 缺少合法的 sampling_rate")
            }
            if (data.spk2id.isNullOrEmpty()) {
                return ConfigParseResult.Failure("config.json 缺少 spk2id，无法确定 speaker")
            }
            ConfigParseResult.Success(config)
        } catch (e: Exception) {
            Log.e(TAG, "parse config failed: $path", e)
            ConfigParseResult.Failure(e.message ?: "config.json 解析失败")
        }
    }

    private const val TAG = "FileUtils"

    @Throws(IOException::class)
    fun copy(src: File?, dst: File?) {
        val `in`: InputStream = FileInputStream(src)
        val out: OutputStream = FileOutputStream(dst)

        // Transfer bytes from in to out
        val buf = ByteArray(1024)
        var len: Int
        while (`in`.read(buf).also { len = it } > 0) {
            out.write(buf, 0, len)
        }
        `in`.close()
        out.close()
    }

    /**
     * Delete a directory and all its contents
     */
    fun deleteDirectory(fileOrDirectory: File) {
        val children = fileOrDirectory.listFiles()
        if (fileOrDirectory.isDirectory && children != null) {
            for (child in children) {
                deleteDirectory(child)
            }
        }
        fileOrDirectory.delete()
    }

    /**
     * Recursively copy the files in the Asset directory to rootDir
     * @param assets
     * @param path
     * @param rootDir
     * @throws IOException
     */
    @Throws(IOException::class)
    fun copyAssets(assets: AssetManager, path: String, rootDir: String, buffer: ByteArray?) {
        if (isAssetsDir(assets, path)) {
            val dir = File(rootDir + File.separator + path)
            check(!(!dir.exists() && !dir.mkdirs())) { "mkdir failed" }
            for (s in assets.list(path)!!) {
                copyAssets(assets, "$path/$s", rootDir, buffer)
            }
        } else {
            val input = assets.open(path)
            val dest = File(rootDir, path)
            copyToFileOrThrow(input, dest, buffer)
        }
    }

    fun isAssetsDir(assets: AssetManager, path: String?): Boolean {
        try {
            val files = assets.list(path!!)
            return files != null && files.size > 0
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return false
    }

    var DEFAULT_BUFFER = 64 * 1024

    @Throws(IOException::class)
    fun copyToFileOrThrow(inputStream: InputStream, destFile: File, buffer: ByteArray?) {
        var buffer = buffer
        if (destFile.exists()) {
            return
        }
        val file = destFile.parentFile
        if (file != null && !file.exists()) {
            file.mkdirs()
        }
        val out = FileOutputStream(destFile)
        try {
            if (buffer == null || buffer.size == 0) {
                buffer = ByteArray(DEFAULT_BUFFER)
            }
            // byte[] buffer = new byte[BUFFER];
            var bytesRead: Int
            while (inputStream.read(buffer).also { bytesRead = it } >= 0) {
                out.write(buffer, 0, bytesRead)
            }
        } finally {
            out.flush()
            try {
                out.fd.sync()
            } catch (e: IOException) {
            }
            out.close()
        }
    }
}

/**
 * [FileUtils.parseConfig] 的结果。失败带上可直接展示给用户的原因。
 */
sealed interface ConfigParseResult {
    data class Success(val config: Bv2Config) : ConfigParseResult
    data class Failure(val reason: String) : ConfigParseResult
}
